atoma-proxy/src/server/middleware.rs

```diff
@@ -687,16 +687,35 @@ pub async fn handle_locked_stack_middleware(
             // We need to acquire a new stack for the request, to be able to retry
             let user_id = request_metadata.user_id;
             let max_total_num_compute_units = request_metadata.max_total_num_compute_units;
-            // 1. Acquire a new stack for the request, this will also lock compute units for the new acquired stack
-            let selected_node_metadata = get_cheapest_node_and_acquire_new_stack(
+            // 1. Try to get a Stack from the state manager
```

Comment:
Do you think there could be a potential race condition here?

```rust
// Use a mutex or other synchronization mechanism
let stack_lock = state.stack_acquisition_mutex.lock().await;
let result = acquire_stack_with_retry(&state, user_id, &request_metadata).await;
drop(stack_lock); // Release the lock
```

Reply:
We have already handled this before. This code should be safe

Reply:
No, we have handled race conditions situations before.

Reply:
Btw, the suggestion would not help in this case, as it would try to always buy stacks, instead of querying the db if a lock already exists on some other task.